scroll_derive/src/lib.rs

```diff
@@ -15,20 +21,63 @@ fn impl_field(ident: &proc_macro2::TokenStream, ty: &syn::Type) -> proc_macro2::
             }) => {
                 let size = int.base10_parse::<usize>().unwrap();
                 quote! {
-                    #ident: { let mut __tmp: #ty = [0u8.into(); #size]; src.gread_inout_with(offset, &mut __tmp, ctx)?; __tmp }
+                    #ident: { let mut __tmp: #ty = [0u8.into(); #size]; src.gread_inout_with(offset, &mut __tmp, #ctx)?; __tmp }
                 }
             }
             _ => panic!("Pread derive with bad array constexpr"),
         },
-        syn::Type::Group(ref group) => impl_field(ident, &group.elem),
+        syn::Type::Group(ref group) => impl_field(ident, &group.elem, custom_ctx),
         _ => {
             quote! {
-                #ident: src.gread_with::<#ty>(offset, ctx)?
+                #ident: src.gread_with::<#ty>(offset, #ctx)?
             }
         }
     }
 }
 
+/// Retrieve the field attribute with given ident e.g:
+/// ```ignore
+/// #[attr_ident(..)]
+/// field: T,
+/// ```
+fn get_attr<'a>(attr_ident: &str, field: &'a syn::Field) -> Option<&'a syn::Attribute> {
+    field
+        .attrs
+        .iter()
+        .filter(|attr| attr.path().is_ident(attr_ident))
+        .next()
+}
+
+/// Gets the `TokenStream` for the custom ctx set in the `ctx` attribute. e.g. `expr` in the following
+/// ```ignore
+/// #[scroll(ctx = expr)]
+/// field: T,
+/// ```
+fn custom_ctx(field: &syn::Field) -> Option<proc_macro2::TokenStream> {
+    get_attr("scroll", field).and_then(|x| {
+        // parsed #[scroll..]
+        // `expr` is `None` if the `ctx` key is not used.
+        let mut expr = None;
+        let res = x.parse_nested_meta(|meta| {
+            // parsed #[scroll(..)]
+            if meta.path.is_ident("ctx") {
+                // parsed #[scroll(ctx..)]
+                let value = meta.value()?; // parsed #[scroll(ctx = ..)]
+                expr = Some(value.parse::<syn::Expr>()?); // parsed #[scroll(ctx = expr)]
```

Comment:
is there a reason you don't `return Ok(value.parse::<syn::Expr>()?);` ? this way on the match below you can do `Ok(expr) => expr.into_token_stream()` without the map, which seems slightly more idiomatic to me, but perhaps there's some issue i'm not seeing?

Reply:
`parse_nested_meta` in `syn` has this signature:
```rs
pub fn parse_nested_meta(
        &self,
        logic: impl FnMut(ParseNestedMeta) -> Result<()>,
    ) -> Result<()> {
        self.parse_args_with(meta::parser(logic))
    }
```
So we have to return `Ok(())` on success.